src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/TagHelperDescriptorExtensions.cs

```diff
@@ -9,6 +9,16 @@ namespace Microsoft.AspNetCore.Razor.Language.Components
 {
     internal static class TagHelperDescriptorExtensions
     {
+        public static bool IsAnyComponentDocumentTagHelper(this TagHelperDescriptor tagHelper)
+        {
+            if (tagHelper == null)
+            {
+                throw new ArgumentNullException(nameof(tagHelper));
+            }
+
+            return tagHelper.IsComponentTagHelper() || tagHelper.Metadata.ContainsKey(ComponentMetadata.SpecialKindKey);
```

Comment:
Hmm showerthought, thinking back we should have used the `Kind` property on TagHelperDescriptor to separate functionality.

Reply:
yeah the reason why this doesn't is that there was an ancient bug where `kind` wasn't serialized, and we needed Blazor to work on that VS release.